### ISSUE TITLE
Wrong implementation of tmin?

### ISSUE THREAD

**Issue by wtdcode** [AFLplusplus/LibAFL#1635]
These lines:

https://github.com/AFLplusplus/LibAFL/blob/02cd260af0ca49027e5d4f36aca459836be5d04c/libafl/src/stages/tmin.rs#L124-L131

And these:

https://github.com/AFLplusplus/LibAFL/blob/02cd260af0ca49027e5d4f36aca459836be5d04c/libafl/src/stages/tmin.rs#L152-L156

Indicate that only `base` should be preserved if it's interesting. It looks like the right condition should be `if !feedback.is_interesting(state, manager, &input, observers, &exit_kind)?`?

**Comment by tokatoka** [comment]
yes looks like the opposite
@addisoncrump

**Comment by addisoncrump** [comment]
Explanation comments are bad here, sorry. There are three sets of feedback present in this code. Two sets of feedback are the fuzzer feedbacks, which will add things to the corpus/solutions if an execution is marked as interesting (`fuzzer.feedback_mut()`/`fuzzer.objective_mut()`). The other is the feedback generated to determine if the input we just found is a smaller, but equivalent case (`feedback`) which is determined by the `is_interesting` call ("testcase is interesting because it is equivalent according to user").

The first comment is saying that it will not be considered interesting by the _fuzzer_ feedback because we know for a fact that we added a new corpus entry, i.e. it should not be considered a smaller equivalent feedback because we covered new things with it. We determine if this is the case by recording the corpus and solution corpus size before and checking if it's equal afterward, i.e. "was anything added to the corpora?".

The second comment is saying that, since the fuzzer has not previously marked the input as interesting (a requirement for `base` to be set to that input), then we don't need to check the return value of the `fuzzer.feedback_mut()` -- instead we only use it because some feedbacks build metadata in `is_interesting`, then add it in `add_metadata`. This is unfortunately necessary because we cannot commit the metadata beforehand without consuming the input in a testcase.
